lib/controllers/calendar.dart

```diff
@@ -0,0 +1,258 @@
+        defaultTargetPlatform == TargetPlatform.iOS) {
+      return _MobileCalendarController();
+    }
+    throw UnsupportedError(
+        'CalendarController is not supported on this platform');
+  }
+
+  static CalendarController get instance =>
+      _instance ??= CalendarController._getByPlatform();
+
+  Future<bool> isAvailable();
+  Future<bool> requestPermission();
+
+  Future<void> upsertEvent({
+    required InstanceMember subscription,
+    required Instance instance,
+  });
+  Future<void> editEventRSVP();
+  Future<void> findEvent();
+}
+
+class _MobileCalendarController implements CalendarController {
+  static const defaultCalendarAccountName = "SquadQuest";
+  static const defaultCalendarName = "SquadQuest Events";
+
+  final DeviceCalendarPlugin _calendar = DeviceCalendarPlugin();
+
+  // As an example, our default timezone is UTC.
+  Location _currentLocation = getLocation('UTC');
+
+  _MobileCalendarController() {
+    _checkCurrentLocation();
+  }
+
+  Future _checkCurrentLocation() async {
+    String timezone = 'UTC';
+    try {
+      timezone = await FlutterTimezone.getLocalTimezone();
+    } catch (e) {
+      print('Could not get the local timezone');
+    }
+    _currentLocation = getLocation(timezone);
+    tz.setLocalLocation(_currentLocation);
+  }
+
+  @override
+  Future<bool> isAvailable() async {
+    return (await _calendar.hasPermissions()).data ?? false;
+  }
+
+  @override
+  Future<void> upsertEvent(
+      {required InstanceMember subscription,
+      required Instance instance}) async {
+    await requestPermission();
+
+    final calendarId = await _getCalendar();
+    final existingEvent = await _getEventIdByInstance(calendarId, instance);
+
+    final user = subscription.member;
+    final event = instance;
+
+    final startTime = event.startTimeMin;
+
+    final result = await _calendar.createOrUpdateEvent(Event(
+      calendarId,
+      eventId: existingEvent,
+      title: event.title,
+      status: switch (event.status) {
+        InstanceStatus.draft => null,
+        InstanceStatus.live => EventStatus.Confirmed,
+        InstanceStatus.canceled => EventStatus.Canceled,
+      },
+      description:
+          "${event.notes}\n\nSquadQuest event: https://squadquest.app/event/${subscription.instanceId}",
+      start: TZDateTime.from(startTime, _currentLocation),
+      allDay: event.endTime == null,
+      end: event.endTime != null
+          ? TZDateTime.from(event.endTime!, _currentLocation)
+          : TZDateTime(
+              _currentLocation,
+              startTime.year,
+              startTime.month,
+              startTime.day,
+            ), // besides the nullality, this parameter is required
+      location: event.locationDescription,
+      reminders: [
+        Reminder(minutes: 60),
+      ],
+      attendees: [
+        Attendee(
+          name: user!.fullName,
+          // TODO(@themightychris): Email is mandatory. Do we have a standard email address?
```

Comment:
Hmmm no. Maybe we just use a static placeholder like `calendars@squadquest.app` since at this point we're never going to be putting multiple attendees into an event so they don't need to be unique